// @grammyjs/storage-free/dist/cjs/adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.freeStorage = void 0;
class Storage {
    constructor(token, rootUrl = 'https://grammy-free-session.deno.dev') {
        Object.defineProperty(this, "token", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: token
        });
        Object.defineProperty(this, "rootUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: rootUrl
        });
        Object.defineProperty(this, "jwt", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
    async login() {
        if (this.jwt === undefined) {
            const url = `${this.rootUrl}/login`;
            const body = JSON.stringify({ token: this.token });
            const response = await retryFetch(url, { method: 'POST', body });
            const { token } = await response.json();
            if (typeof token !== 'string') {
                throw new Error('Cannot use free session, invalid bot token!');
            }
            this.jwt = token;
        }
        return this.jwt;
    }
    logout() {
        this.jwt = undefined;
    }
    async call(method, key, body) {
        // perform request
        const url = `${this.rootUrl}/session/${key}`;
        const jwt = await this.login();
        const headers = { 'Authorization': `Bearer ${jwt}` };
        const response = await retryFetch(url, { method, body, headers });
        // handle response
        if (response.status === 401) {
            // token was revoked, must login again
            this.logout();
            return await this.call(method, key, body);
        }
        else if (response.status === 404) {
            // empty session
            return undefined;
        }
        else if (200 <= response.status && response.status < 300) {
            // success
            return method === 'GET' ? await response.text() : undefined;
        }
        else {
            // error
            throw new Error(`${response.status}: ${(await response.json()).error}`);
        }
    }
}
/**
 * @param token The bot token of your bot.
 * @param opts Further configuration options
 * @returns An adapter to grammY's free session storage
 */
function freeStorage(token, opts) {
    const storage = new Storage(token, opts === null || opts === void 0 ? void 0 : opts.rootUrl);
    if ((opts === null || opts === void 0 ? void 0 : opts.jwt) !== undefined)
        storage.jwt = opts.jwt;
    return {
        async read(key) {
            const session = await storage.call('GET', key);
            return session === undefined ? undefined : JSON.parse(session);
        },
        async write(key, data) {
            await storage.call('POST', key, JSON.stringify(data));
        },
        async delete(key) {
            await storage.call('DELETE', key);
        },
        /**
         * Returns the storage authentication token which is used to store the
         * session data. Only useful if you want to avoid the login call that will
         * be performed automatically when the storage adapter contacts its backend
         * for the first time. This can improve startup performance and is
         * especially useful in serverless environments.
         */
        async getToken() {
            return await storage.login();
        },
    };
}
exports.freeStorage = freeStorage;
async function retryFetch(...args) {
    let res;
    do {
        res = await (0, shim_node_js_1.fetch)(...args);
        if (res.status >= 500) {
            console.error(`${res.status} in free session service, retrying!`);
        }
    } while (res.status >= 500);
    return res;
}
const shim_node_js_1 = require("./shim.node.js");
